File: Leet Code Solutions/1420. Build Array Where You Can Find The Maximum Exactly K Comparisons/PralinKhaira-K-Comparison.py
class Solution:
    def numOfArrays(self, n: int, m: int, k: int) -> int:
        mod = 10**9 + 7
        
        dp = [[0] * (k+1) for _ in range(m+1)]
        prefix = [[0] * (k+1) for _ in range(m+1)]
        prevDp = [[0] * (k+1) for _ in range(m+1)]
        prevPrefix = [[0] * (k+1) for _ in range(m+1)]
        
        for j in range(1, m+1):
            prevDp[j][1] = 1
            prevPrefix[j][1] = j
        
        for _ in range(2, n+1):
            dp = [[0] * (k+1) for _ in range(m+1)]
            prefix = [[0] * (k+1) for _ in range(m+1)]
            
            for maxNum in range(1, m+1):
                for cost in range(1, k+1):
                    dp[maxNum][cost] = (maxNum * prevDp[maxNum][cost]) % mod
                    
                    if maxNum > 1 and cost > 1:
                        dp[maxNum][cost] += prevPrefix[maxNum - 1][cost - 1]
                        dp[maxNum][cost] %= mod
                    
                    prefix[maxNum][cost] = (prefix[maxNum - 1][cost] + dp[maxNum][cost]) % mod
            
            prevDp, prevPrefix = [row[:] for row in dp], [row[:] for row in prefix]
        
        return prefix[m][k]

File: Leet Code Solutions/1420. Build Array Where You Can Find The Maximum Exactly K Comparisons/PralinKhaira-K-Comparison.rs
impl Solution {
    pub fn num_of_arrays(n: i32, m: i32, k: i32) -> i32 {
        const MOD: i32 = 1_000_000_007;

        let mut dp = vec![vec![0; (k + 1) as usize]; (m + 1) as usize];
        let mut prefix = vec![vec![0; (k + 1) as usize]; (m + 1) as usize];
        let mut prevDp = vec![vec![0; (k + 1) as usize]; (m + 1) as usize];
        let mut prevPrefix = vec![vec![0; (k + 1) as usize]; (m + 1) as usize];

        for j in 1..=m {
            prevDp[j as usize][1] = 1;
            prevPrefix[j as usize][1] = j;
        }

        for _ in 2..=n {
            for maxNum in 1..=m {
                for cost in 1..=k {
                    dp[maxNum as usize][cost as usize] = 
                        ((maxNum as i64 * prevDp[maxNum as usize][cost as usize] as i64) % MOD as i64) as i32;
                    
                    if maxNum > 1 && cost > 1 {
                        dp[maxNum as usize][cost as usize] = 
                            (dp[maxNum as usize][cost as usize] + prevPrefix[(maxNum-1) as usize][(cost-1) as usize]) % MOD;
                    }

                    prefix[maxNum as usize][cost as usize] = 
                        (prefix[(maxNum-1) as usize][cost as usize] + dp[maxNum as usize][cost as usize]) % MOD;
                }
            }

            prevDp = dp.clone();
            prevPrefix = prefix.clone();
        }

        prefix[m as usize][k as usize]
    }
}